src/Lecture2.hs

```diff
@@ -152,12 +183,124 @@ You're free to define any helper functions.
 
 -- some help in the beginning ;)
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
+    { knightHealth    :: Health
+    , knightAttack    :: Attack
+    , knightEndurance :: Endurance
+    } deriving (Eq)
 
-dragonFight = error "TODO"
+data Dragon = Dragon
+    { dragonHealth    :: Health
+    , dragonFirePower :: FirePower
+    } deriving (Eq)
+
+data Chest a = Chest
+    { gold      :: Gold
+    , threasure :: Maybe a
+    } deriving (Show, Eq)
+
+data Reward = Reward Gold (Maybe Treasure) XP deriving (Show, Eq)
+
+data Treasure = Gems | Jewels deriving (Show, Eq)
+
+data EvilDragon = Green Dragon Gold
+                | Black Dragon (Chest Treasure)
+                | Red   Dragon (Chest Treasure) deriving (Eq)
```

Comment:
I see, I think I get that idea from here:

* A chest contains a non-zero amount of gold and a **possible** treasure.

I thought that a Chest may contain a Treasure.

Reply:
It's "possible" treasure because not every dragon contains treasure 🙂 

Your implementation is already good enough 👍🏻 
I was just saying that it's possible to enforce this condition even more.